### DISCUSSION TITLE
Upgrading a react typescript project from defencedigital 2.80 to 3.30

### DISCUSSION THREAD

**Discussion by RichardP932** [Royal-Navy/design-system#3164] [General]
**Upgrading a react typescript project from defencedigital 2.80 to 3.30** 

Hi, 

All help greately appreciated.

If I can sort all these out an use the latest version it will be very nice. If not I will have to abort the branch and carry on with 2.80.

Best regards,
Richard

I have these in my package json after upgrading from 2.80

   "@defencedigital/css-framework": "^2.80.3",
    "@defencedigital/design-tokens": "^3.2.2",
    "@defencedigital/eslint-config-react": "^3.2.2",
    "@defencedigital/fonts": "^3.2.2",
    "@defencedigital/icon-library": "^3.3.0",
    "@defencedigital/react-component-library": "^3.2.2",

I have knocked these out of the devDependencies
    "@types/react-pdf": "^5.0.9",
    "eslint": "^8.8.0",
    "eslint-plugin-prettier": "^4.0.0",

React is;
    "@types/react": "^17.0.40",
    "react": "^17.0.2",
    "react-scripts": "^5.0.0",
    "styled-components": "^5.3.3",

npm Install gives no errors.

npm start gives errors as follows;
export 'Tab' (imported as 'Tab') was not found in '@defencedigital/react-component-library'
export 'BUTTON_SIZE' (imported as 'BUTTON_SIZE') was not found in '@defencedigital/react-component-library'

TS2305: Module '"@defencedigital/react-component-library"' has no exported member 'Tab'.
    2 | import {
    3 |   TabSet,
  > 4 |   Tab,
      |   ^^^
    5 |   BreadcrumbsItem,
    6 |   Breadcrumbs,
    7 | } from '@defencedigital/react-component-library'

ERROR in src/pages/LotMitigation/PerformanceMeasureCard.tsx:5:3

TS2305: Module '"@defencedigital/react-component-library"' has no exported member 'BUTTON_SIZE'.
    3 |   Button,
    4 |   BUTTON_VARIANT,
  > 5 |   BUTTON_SIZE,
      |   ^^^^^^^^^^^
    6 | } from '@defencedigital/react-component-library'

I have other errors but I think that they are to do with upgrading esLint, a few like this;
TS2769: No overload matches this call.
  Overload 1 of 2, '(props: (({ children?: ((string | number | boolean | {} | ReactElement<any, string | JSXElementConstructor<any>> | Iterable<ReactNode> | ReactPortal | ReactNode[]) & (boolean | ... 2 more ... | ReactPortal)) | null | undefined; ... 18 more ...; footnote?: string | undefined; } | { ...; } | { ...; }) & { ...; }) & { ...; }): ReactElement<...>', gave the following error.
    Type '(event: React.ChangeEvent<HTMLInputElement>) => void' is not assignable to type '(event: ChangeEvent<HTMLInputElement> | MouseEvent<HTMLButtonElement, MouseEvent>, newValue: number | null) => void'.
      Types of parameters 'event' and 'event' are incompatible.
        Type 'ChangeEvent<HTMLInputElement> | MouseEvent<HTMLButtonElement, MouseEvent>' is not assignable to type 'ChangeEvent<HTMLInputElement>'.
          Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is not assignable to type 'ChangeEvent<HTMLInputElement>'.
            Types of property 'target' are incompatible.
              Type 'EventTarget' is not assignable to type 'EventTarget & HTMLInputElement'.
                Type 'EventTarget' is missing the following properties from type 'HTMLInputElement': accept, align, alt, autocomplete, and 328 more.
  Overload 2 of 2, '(props: StyledComponentPropsWithAs<FC<NumberInputProps>, any, {}, never, FC<NumberInputProps>, FC<NumberInputProps>>): ReactElement<...>', gave the following error.
    Type '(event: React.ChangeEvent<HTMLInputElement>) => void' is not assignable to type '(event: ChangeEvent<HTMLInputElement> | MouseEvent<HTMLButtonElement, MouseEvent>, newValue: number | null) => void'.
    251 |               <ScoreNumberInput
    252 |                 name="currentValue"
  > 253 |                 onChange={onChange}
        |                 ^^^^^^^^
    254 |                 onBlur={onBlur}

**Comment by RichardP932** [comment]
Just renaming Tab to TabSet has sorted one of the errors. Now working on the BUTTON_SIZE one.

**Comment by RichardP932** [comment]
That was easy too just change the constant no longer exists just use size="small"

**Comment by RichardP932** [comment]
ChangeEvent one looks tricky I have code that does this.
const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const {
      target: { name, value },
    } = event

    const newFormData = {
      ...formData,
      [name]: value,
    }

**Reply by jpveooys** [reply]
Hi @RichardP932 :wave: 

Is that `onChange` error related to the `NumberInput`? If so the type of `onChange` is:

```
  (
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.MouseEvent<HTMLButtonElement>,
    newValue: number | null
  ) => void
```

You'll need to update your callback to match, and I'd use `newValue` rather than `event.target.value` as `event.target` could be one of the stepper buttons.

If you haven't already, I'd recommend having a look at the [migration guide](https://design-system.digital.mod.uk/guidance/migrating-to-v3) as it may help with any other problems you encounter.

**Reply by RichardP932** [reply]
Thank you that is very helpful.

I've moved on a bit but for a while now I have been trying to figure out what the syntax of the callback parameters are.

BEFORE

const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const {
      target: { name, value },
    } = event

    const newFormData = {
      ...formData,
      [name]: value,
    }

    setLastSaved(null)
    updateFormData(newFormData)
  }

AFTER
const onChange = (
  event: React.ChangeEvent<HTMLInputElement> | 
  React.MouseEvent<HTMLButtonElement>,
  newValue: number | null
) => {
  const {
    target: { **name, value** },
  } = event

    const newFormData = {
      ...formData,
      [name]: value,
    }

    setLastSaved(null)
    updateFormData(newFormData)
  }

I have emboldened the errors. I can't figure out how I am to use newValue instead of name, value?

**Reply by RichardP932** [reply]
@jpveooys The migration guide offers a lot less information than you have posted here as does the StoryBook help. There is no example.

**Reply by jpveooys** [reply]
Hi @RichardP932 

> I can't figure out how I am to use newValue instead of name, value?

I'd remove:

```
const {
target: { name, value },
} = event
```

completely.

`newValue` would be used in place of `value` where you are setting `newFormData` further down.

That leaves `name` – but normally that would be already known from the context (i.e. the field the callback is bound to).

We have some form examples here which you may find useful: https://github.com/defencedigital/mod-uk-design-system/tree/master/packages/react-component-library/src/forms

For the Formik example the NumberInput callback is defined as below:

https://github.com/defencedigital/mod-uk-design-system/blob/4a5efd8d24317c6cdf89932f5fb9fa561cc82fc3/packages/react-component-library/src/forms/formik/formik.stories.tsx#L200-L207

**Reply by RichardP932** [reply]
Thank you very much @jpveooys that seems to have got me past the issue.

Within the page the onChange method is used by both the NumberInput and the TextInput controls so it complains further down in the typescript most probably becuse newValue: number | null is not text so it should be a string. Would you recommend changing the method to take both number and string or go with an onChangeNumber, onChangeString?

**Reply by jpveooys** [reply]
Hi @RichardP932 

I'd go with the latter – you'll probably want to stick with the original approach for `TextInput` as the callback is typed as `(event: React.ChangeEvent<HTMLInputElement>) => void` there.

(Though I'd use `event.currentTarget` instead of `event.target` if you can. `event.currentTarget` will be typed as an `HTMLInputElement` for an event of type `React.ChangeEvent<HTMLInputElement>`. `event.target` could be something else in other scenarios if the event has bubbled.)

**Reply by RichardP932** [reply]
Really sorry @jpveooys I am still struggling with the text one which I thought would be fairly easy.
I have method
  const onChangeTextInput = (
    event: React.ChangeEvent<HTMLInputElement>,
    newValue: string | null
    ) => {
  
      const newFormData = {
        ...formData,
        newValue,
      }
      setLastSaved(null)
      updateFormData(newFormData)
    }

Called from this
            <JustificationInput
              name="justification"
              value={formData.justification || ''}
              **onChange={onChangeTextInput}**
              onBlur={onBlur}
              isDisabled={!hasPostMitigationScore}
              isInvalid={Boolean(errors?.justification)}
            />
I have made bold where I get the error appear;
(property) onChange?: ((event: React.ChangeEvent<HTMLInputElement>) => void) | undefined
Optional handler invoked when the value of the component changes.

No overload matches this call.
  Overload 1 of 2, '(props: { label: string; type?: TextInputType | undefined; name: string; ref?: Ref<HTMLInputElement> | undefined; key?: Key | null | undefined; ... 12 more ...; startAdornment?: ReactNode; } & { ...; } & { ...; }): ReactElement<...>', gave the following error.
    Type '(event: React.ChangeEvent<HTMLInputElement>, newValue: string | null) => void' is not assignable to type '(event: ChangeEvent<HTMLInputElement>) => void'.
  Overload 2 of 2, '(props: StyledComponentPropsWithAs<ForwardRefExoticComponent<TextInputProps & RefAttributes<HTMLInputElement>>, any, TextInputProps, never, ForwardRefExoticComponent<...>, ForwardRefExoticComponent<...>>): ReactElement<...>', gave the following error.
    Type '(event: React.ChangeEvent<HTMLInputElement>, newValue: string | null) => void' is not assignable to type '(event: ChangeEvent<HTMLInputElement>) => void'.ts(2769)
[TextInput.d.ts(41, 5): ]()The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & { label: string; type?: TextInputType | undefined; name: string; ref?: Ref<HTMLInputElement> | undefined; ... 13 more ...; startAdornment?: ReactNode; } & { ...; } & { ...; }'
[TextInput.d.ts(41, 5): ]()The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & { label: string; type?: TextInputType | undefined; name: string; ref?: Ref<HTMLInputElement> | undefined; ... 13 more ...; startAdornment?: ReactNode; } & { ...; } & { ...; }'

**Reply by RichardP932** [reply]
I think that React.InputHTMLAttributes<HTMLInputElement> defines onChange as (event: ChangeEvent<HTMLInputElement>) => void), and in onChange it is defined as ChangeHandler<string>.

Because it is typescript it is quite strict on types.

**Reply by jpveooys** [reply]
Hi @RichardP932 

`TextInput` is not special like `NumberInput` and so its `onChange` doesn't have a `newValue` argument. So you can use `event.currentTarget.value` for the value in that one.

(Make sure it's assigned to the right property in formData too – looks like it would be `justification` from the below.)

**Reply by RichardP932** [reply]
Thank you @jpveooys that worked.

**Comment by RichardP932** [comment]
The type that is causing the problem is value={formData.justification || ''} which is defined in the class as

interface FormData {
  currentValue: number | null
  noData: boolean
  justification: string | null
}
